### PR TITLE
Dev edition

### PR DESCRIPTION
**Pull request by ochameau**

### CODE REVIEW

**Comment by ochameau** [comment]
@jryans Could you take a brief look at that patch to be able to give your opinion on this?
I would like to discuss about that in the next devtools-go-faster meeting.

You can test this pull request by installing this addon in all-but-release-Firefox as this isn't signed:
http://techno-barje.fr/github/pr/8f3e764b66312edb089c2c1446a13c029bae2590/devtools.xpi
And so with xpinstall.signatures.required set to false...

It also demonstrates how easy it is to test a devtools pull request via github and the addon workflow...
This xpi is automatically built. You can find the link to it by clicking on "show all checks" and the "details" link next to "default  -- Add-on" item.

**Comment by ochameau** [comment]
Note that for now, you may have to press Ctrl+Alt+R after addon install in order to reload the devtools from the addon... And do not forget to uninstall the addon after testing as it will persist over reboots!

File: client/menus.js [hunk 1]
```diff
@@ -192,4 +192,149 @@ exports.menuitems = [
...
+      let profileService = Cc["@mozilla.org/toolkit/profile-service;1"]
+        .getService(Ci.nsIToolkitProfileService);
+
+      let profile;
+      try {
+        // getProfileByName throws if it doesn't exists yet
+        profile = profileService.getProfileByName(profileName);
+      } catch(e) {}
+      if (!profile) {
+        profile = profileService.createProfile(null, profileName);
```
Comment (jryans) [inline client/menus.js:207]: We could also customize it further by setting certain prefs in the new profile if we wanted (similar to how we set up the Browser Toolbox profile).

File: client/menus.js [hunk 2]
```diff
@@ -192,4 +192,149 @@ exports.menuitems = [
...
+        app.moveTo(applications, app.leafName);
+
+        // On latest Mac only signed application can be opened
+        // so ask user to manually flag security exception
+        window.alert("Right click on '"+appName+"' to select 'Open' and accept the security exception.");
+        app.reveal();
+
+        // Return to prevent automatic opening of dev edition on Mac
+        return;
+      } else {
```
Comment (jryans) [inline client/menus.js:294]: Maybe this should be `else if` and test for Linux?

File: client/menus.js [hunk 3]
```diff
@@ -192,4 +192,149 @@ exports.menuitems = [
...
+
+        // Copy the shortcut to Desktop and Start menu
+        let desktop = Services.dirsvc.get("Desk", Ci.nsIFile);
+        let progs = Services.dirsvc.get("Progs", Ci.nsIFile);
+        shortcut.copyTo(desktop, shortcut.leafName);
+        shortcut.copyTo(progs, shortcut.leafName);
+
+        // Remove the temporary shortcut file
+        shortcut.remove(false);
+      } else if (Services.appinfo.OS == "Darwin") {
```
Comment (jryans) [inline client/menus.js:248]: There might be a simpler approach for macOS, such as a shell script using `open -a <APP> --args <PROFILE>`, and some of them might get around the signing issues. Also, I think by default `/Applications` won't be writable by the current user. You could avoid this by writing to `~/Applications`, which is a standard per-user app directory.

There is a (somewhat old) [SO discussion][1] about ways to do this.

[1]: http://superuser.com/questions/16750/how-can-i-run-an-application-with-command-line-arguments-in-mac-os

**Review by jryans — COMMENTED** [review]
I feel like we can find ways to make this work technically with an acceptable risk level. I am wondering more about the UX here. 

With existing Dev. Edition builds, you get the separate profile by default with no user action. With the approach here, someone has to have Firefox already, and then take some UI action to trigger this process of creating a new profile.

How would we message this to users without confusing people but still be discoverable? I guess developers would generally understand language like "Allow Firefox Developer Edition to run side by side with this browser", but I am still not sure where you place that so that it would even be found... Maybe we could use an approach like Dev. Edition promo (it's an overlay above the toolbox) that appears if you open the toolbox in Beta currently? So, on first open of DevTools, offer this separate profile mode.

**Comment by jryans** [comment]
As for the add-on being built in the PR, I think the idea is great! It will make it much easier to try out the changes in the PR.

However, the current version doesn't seem to work for me. I am using a local Nightly build and `xpinstall.signatures.required` is already false. If I just paste in the URL, the install is blocked with no allow button. Does it need to be an HTTPS URL?

I was able to install by using `Save link as...` to download the file first, and then loading it as a temporary add-on in `about:debugging`. I was able to open a toolbox which I think came from the add-on, but then it failed to completely destroy:

```
Trying to register about:devtools-toolbox more than once.  about-devtools-toolbox.js:46
TypeError: 'getInterface' called on an object that does not implement interface Window.  utils.js:19:10
"Warning: Each child in an array or iterator should have a unique "key" prop. Check the render method of `Frame`. See https://fb.me/react-warning-keys for more information.
    in span (created by Frame)
    in Frame (created by Message)
    in span (created by Message)
    in span (created by Message)
    in span (created by Message)
    in div (created by Message)
    in Message (created by ConsoleApiCall)
    in ConsoleApiCall (created by MessageContainer)
    in MessageContainer (created by ConsoleOutput)
    in div (created by ConsoleOutput)
    in ConsoleOutput (created by Connect(ConsoleOutput))
    in Connect(ConsoleOutput)
    in div
    in Provider"  react-dev.js:22807:9
TypeError: CommandUtils.getCommandbarSpec is not a function
Stack trace:
getToolbarSpec@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/framework/toolbox.js:1180:16
Toolbox.prototype._buildButtons<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/framework/toolbox.js:1051:20
_run@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/task.js:311:39
  Promise-backend.js:925
TypeError: this.jsterm is null[Learn More]
```

Anyway, hopefully these are solvable issues! I think the overall idea is great.

**Comment by ochameau** [comment]
3 INFO TEST-OK | client/framework/test/browser_target_events.js | took 737ms
10 INFO TEST-UNEXPECTED-FAIL | client/framework/test/browser_target_from_url.js | undefined assertion name - Got targetFromURL, unsupported type 'x' parameter, expected targetFromURL, unsupported type='x' parameter
24 INFO TEST-UNEXPECTED-FAIL | client/framework/test/browser_target_from_url.js | undefined assertion name - Got targetFromURL, tab with outerWindowID '10000' doesn't exist, expected targetFromURL, tab with outerWindowID:'10000' doesn't exist
62 INFO TEST-OK | client/framework/test/browser_target_from_url.js | took 1070ms
65 INFO TEST-OK | client/framework/test/browser_target_remote.js | took 77ms
68 INFO TEST-OK | client/framework/test/browser_target_support.js | took 111ms
	Passed: 58
	Failed: 2

**Comment by ochameau** [comment]
3 INFO TEST-OK | client/framework/test/browser_target_events.js | took 294ms
10 INFO TEST-UNEXPECTED-FAIL | client/framework/test/browser_target_from_url.js | undefined assertion name - Got targetFromURL, unsupported type 'x' parameter, expected targetFromURL, unsupported type='x' parameter
24 INFO TEST-UNEXPECTED-FAIL | client/framework/test/browser_target_from_url.js | undefined assertion name - Got targetFromURL, tab with outerWindowID '10000' doesn't exist, expected targetFromURL, tab with outerWindowID:'10000' doesn't exist
62 INFO TEST-OK | client/framework/test/browser_target_from_url.js | took 1179ms
65 INFO TEST-OK | client/framework/test/browser_target_remote.js | took 79ms
68 INFO TEST-OK | client/framework/test/browser_target_support.js | took 124ms
	Passed: 58
	Failed: 2

**Pull request closed without merging**